File: Polars.FSharp.Tests/ExcelTests.fs
namespace Polars.FSharp.Tests

open System
open System.IO
open Xunit
open Polars.FSharp
open MiniExcelLibs // 这是一个 .NET 库，F# 调用完全没问题

// 定义一个 Record Type，MiniExcel 会自动将其属性映射为 Excel 列
type TestExcelRow = {
    Id: int
    Name: string
    Score: float
    IsActive: bool
    JoinDate: DateTime
}

type FSharpExcelTests() =

    [<Fact>]
    member _.``IO: Read Excel (Native Roundtrip with MiniExcel)`` () =
        let tempFile = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString() + ".xlsx")
        
        try
            // ---------------------------------------------------
            // 1. 准备数据 (F# List of Records)
            // ---------------------------------------------------
            let data = [
                { Id = 1; Name = "Alice"; Score = 99.5; IsActive = true;  JoinDate = DateTime(2022, 1, 1) }
                { Id = 2; Name = "Bob";   Score = 88.0; IsActive = false; JoinDate = DateTime(2023, 5, 20) }
                { Id = 3; Name = "C";     Score = 60.0; IsActive = true;  JoinDate = DateTime(2024, 12, 31) }
            ]

            // MiniExcel 支持 IEnumerable，F# List 实现了它
            MiniExcel.SaveAs(tempFile, data) |> ignore
            // =================================================================
            // 2. 测试默认读取
            // =================================================================
            let testDefaultRead() =
                use df = DataFrame.ReadExcel tempFile
                
                Assert.Equal(3L, df.Rows)
                
                // 验证 F# Extension Methods 读取值
                // 注意：Excel 数字默认是 f64 (Double)
                // 验证 Id: 1 -> 1.0 (Polars 默认推断)
                Assert.Equal(1.0, df.Float("Id", 0).Value) 
                
                // 验证 String
                Assert.Equal("Alice", df.String("Name", 0).Value)
                
                // 验证 DateTime (Rust 解析 OADate -> Polars Datetime)
                Assert.Equal(DateTime(2022, 1, 1), df.DateTime("JoinDate", 0).Value)

            testDefaultRead()

            // =================================================================
            // 3. 测试 Schema Overrides (强类型控制)
            // =================================================================
            let testSchemaRead() =
                // 场景：我希望 Id 被读作 Int64，而不是默认的 Float64
                // 场景：我希望 Score 被读作 String
                use schema = PolarsSchema.ofList([
                    "Id", DataType.Int64
                    "Score", DataType.String
                ])
                use df = DataFrame.ReadExcel(tempFile, schema=schema)

                // 验证类型是否生效
                Assert.Equal(DataType.Int64, df.Schema.["Id"])
                Assert.Equal(DataType.String, df.Schema.["Score"])

                // 验证值转换逻辑 (Rust 层 match 逻辑)
                // Id: Float(1.0) -> Int64(1)
                Assert.Equal(1L, df.Int("Id", 0).Value)
                
                // Score: Float(99.5) -> String("99.5")
                Assert.Equal("99.5", df.String("Score", 0).Value)

            testSchemaRead()

        finally
            if File.Exists(tempFile) then File.Delete(tempFile)
    [<Fact>]
    member _.``IO: Excel Roundtrip (Precision & Formats)`` () =
        let tempFile = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString() + ".xlsx")
        
        try
            // 1. 构造极端数据
            // 使用 Series 构造器创建 DataFrame
            // F# 中 uint64 对应 System.UInt64 (Rust u64)
            // 这是一个远超 Excel 53-bit 精度的数
            let bigId = 18446744073709551615UL 
            
            use sId = Series.create("Id", [| 1 |])
            use sBigId = Series.create("BigId", [| bigId |])
            use sDate = Series.create("MyDate", [| DateTime(2023, 10, 1) |])
            
            use df = DataFrame.create [sId; sBigId; sDate]

            // 2. 写入 Excel (带格式)
            // 我们希望 BigId 被自动转为文本，MyDate 按照 dd-mm-yyyy 显示
            df.WriteExcel(
                tempFile, 
                sheetName="Data", 
                dateFormat="dd-mm-yyyy"
            )

            // 3. 读回 Excel (带 Schema 校验)
            // 强制 BigId 为 String，看看它是否保持了原样
            use schema = PolarsSchema.ofList ["BigId", DataType.String]

            use dfRead = DataFrame.ReadExcel(tempFile, schema=schema)

            // 4. 验证
            // BigId 应该是 String 类型，且内容完全一致 (没有被 Excel 截断为 ...1615 -> ...0000)
            Assert.Equal(DataType.String, dfRead.Schema.["BigId"])
            
            // 获取值 (注意：F# 获取 String Series 的值)
            let readBigIdStr = dfRead.String("BigId", 0).Value
            Assert.Equal(bigId.ToString(), readBigIdStr)

            // 验证日期值 (逻辑值应该是对的)
            let readDate = dfRead.DateTime("MyDate", 0).Value
            Assert.Equal(DateTime(2023, 10, 1), readDate)

        finally
            if File.Exists(tempFile) then File.Delete(tempFile)

File: Polars.CSharp.Tests/ExcelTests.cs
#nullable enable
using MiniExcelLibs;

namespace Polars.CSharp.Tests;

public class ExcelTests
{
    [Fact]
    public void Test_ReadExcel_RoundTrip_With_MiniExcel()
    {
        // ---------------------------------------------------
        // 1. 准备数据 (出题)
        // ---------------------------------------------------
        var tempFile = Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid()}.xlsx");

        var data = new[]
        {
            // 混合数据类型，测试推断能力
            new { Id = 1, Name = "Alice", JoinDate = new DateTime(2022, 1, 1), Score = 99.5, IsActive = true,  Phone = 13800138000 },
            new { Id = 2, Name = "Bob",   JoinDate = new DateTime(2023, 5, 20),Score = 88.0, IsActive = false, Phone = 13900139000 },
            new { Id = 3, Name = "Charlie",JoinDate= new DateTime(2024, 12, 31),Score= 60.0, IsActive = true,  Phone = 13700137000 },
        };

        try
        {
            // 用 MiniExcel 生成标准的 .xlsx 文件
            MiniExcel.SaveAs(tempFile, data);

            // =================================================================
            // 2. 测试默认读取 (自然推断)
            // =================================================================
            {
                using var df = DataFrame.ReadExcel(tempFile);
                // df.Show();

                Assert.Equal(3, df.Height);
                
                // 验证日期 (Rust OADate -> Polars Datetime)
                // 注意：Polars 默认 Datetime 是微秒级，C# DateTime 也能对应上
                var dateVal = df.GetValue<DateTime>(0, "JoinDate");
                Assert.Equal(new DateTime(2022, 1, 1), dateVal);

                // 验证数值
                Assert.Equal(99.5, df.GetValue<double>(0, "Score"));
                
                // 验证布尔
                Assert.True(df.GetValue<bool>(0, "IsActive"));
                
                // 验证 Phone (MiniExcel 默认写入为数字，Polars 自然推断为 Int64)
                Assert.Equal(13800138000, df.GetValue<double>(0, "Phone"));
            }

            // =================================================================
            // 3. 测试 Schema Overrides (强转测试)
            // =================================================================
            // 场景：电话号码在 Excel 里存的是数字，但我们需要它作为 String 读入
            {
                // 定义 Schema：强制 Phone 列为 String
                using var schema = new PolarsSchema();
                schema.Add("Phone", DataType.String); 

                using var df = DataFrame.ReadExcel(tempFile, schema: schema);

                // 验证 Phone 现在是不是 String 类型
                Assert.Equal(DataType.String, df.Schema["Phone"]);
                
                // 验证内容是否正确转成了字符串
                Assert.Equal("13800138000", df.GetValue<string>(0, "Phone"));
                
                // 验证其他未指定的列是否依然正常推断 (Score 应该是 Float64)
                Assert.Equal(DataType.Float64, df.Schema["Score"]);
            }
        }
        finally
        {
            if (File.Exists(tempFile)) File.Delete(tempFile);
        }
    }

    [Fact]
    public void Test_ReadExcel_EmptyRows_And_Nulls()
    {
        var tempFile = Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid()}.xlsx");

        // 构造带空行和 Null 的数据
        // MiniExcel 写 null 会产生空单元格
        var data = new[]
        {
            new { Col1 = (string?)"A", Col2 = (int?)1 },
            new { Col1 = (string?)null, Col2 = (int?)null },
            new { Col1 = (string?)"C", Col2 = (int?)3 },
        };

        try
        {
            MiniExcel.SaveAs(tempFile, data);

            // 测试 dropEmptyRows = true (默认)
            using var dfDropped = DataFrame.ReadExcel(tempFile, dropEmptyRows: true);
            Assert.Equal(2, dfDropped.Height); // 中间那行应该没了
            Assert.Equal("A", dfDropped.GetValue<string>(0, "Col1"));
            Assert.Equal("C", dfDropped.GetValue<string>(1, "Col1"));

            // 测试 dropEmptyRows = false
            using var dfKept = DataFrame.ReadExcel(tempFile, dropEmptyRows: false);
            Assert.Equal(3, dfKept.Height);
            Assert.Null(dfKept.GetValue<string>(1, "Col1")); // 第二行存在，且为 Null
        }
        finally
        {
            if (File.Exists(tempFile)) File.Delete(tempFile);
        }
    }
    [Fact]
    public void Test_WriteExcel_Precision_And_Formats()
    {
        var tempFile = Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid()}.xlsx");

        try
        {
            // 1. 构造极其刁钻的数据
            // UInt64.MaxValue 是 18446744073709551615，远超 Excel 精度
            // Date 和 Datetime 用于测试格式化
            using var df = DataFrame.From(
            [
                new 
                { 
                    Id = 1, 
                    BigId = ulong.MaxValue, // 极端大数
                    JoinDate = new DateTime(2023, 10, 1),
                    LogTime = new DateTime(2023, 10, 1, 14, 30, 0)
                }
            ]);

            // 2. 写入 Excel (自定义格式)
            df.WriteExcel(
                tempFile, 
                sheetName: "Data", 
                dateFormat: "dd/mm/yyyy",          // 英国格式
                datetimeFormat: "hh:mm AM/PM"      // 12小时制
            );

            // 3. 验证 (这里我们“回读”来验证)
            // 注意：回读时，BigId 应该是 String 类型，因为我们在 Write 时强转了
            using var dfRead = DataFrame.ReadExcel(
                tempFile, 
                schema: new PolarsSchema().Add("BigId", DataType.String) // 强制按 String 读回以验证内容
            );

            // 验证大数精度：如果转成了数字，末尾肯定不是 ...1615
            // 如果转成了 String，应该完全一致
            var bigIdStr = dfRead.GetValue<string>(0, "BigId");
            Assert.Equal(ulong.MaxValue.ToString(), bigIdStr);

            // 验证日期值 (格式化只影响显示，ReadExcel 读回来的应该是正确的 DateTime 值)
            var joinDate = dfRead.GetValue<DateTime>(0, "JoinDate");
            Assert.Equal(new DateTime(2023, 10, 1), joinDate);
        }
        finally
        {
            if (File.Exists(tempFile)) File.Delete(tempFile);
        }
    }
}